#Exercise 1: Handle duplicates in Breast Cancer dataset
In this exercise, we will learn how to identify duplicate records and how to handle this kind of issues.
The dataset used for this exercise is the Breast Cancer Detection shared by KDD: https://www.kdd.org/kdd-cup/view/kdd-cup-2008/Data

1. Open on a new Colab notebook and import the pandas package

In [0]:
import pandas as pd

2. Assign the link to the Breast Cancer dataset to a variable called 'file_url':

In [0]:
file_url = 'https://raw.githubusercontent.com/TrainingByPackt/The-Data-Science-Workshop/master/Chapter11/breast-cancer-wisconsin.data'

3. Using the read_csv method from the package pandas, load the dataset into a new variable called 'df' with the parameter header=None as this file doesn't contain column names:

In [0]:
df = pd.read_csv(file_url, header=None)

4. Create a variable called col_names containing the names of the columns: 'Sample code number','Clump Thickness','Uniformity of Cell Size','Uniformity of Cell Shape','Marginal Adhesion','Single Epithelial Cell Size',
'Bare Nuclei','Bland Chromatin','Normal Nucleoli','Mitoses','Class' (they are specified in this file https://archive.ics.uci.edu/ml/machine-learning-databases/breast-cancer-wisconsin/breast-cancer-wisconsin.names)

In [0]:
col_names = ['Sample code number','Clump Thickness','Uniformity of Cell Size','Uniformity of Cell Shape','Marginal Adhesion','Single Epithelial Cell Size',
'Bare Nuclei','Bland Chromatin','Normal Nucleoli','Mitoses','Class']                  

5. Assign the column names of the dataframe using the attribute 'columns':

In [0]:
df.columns = col_names

6. Display the shape of the dataframe using the attribute '.shape':

In [6]:
df.shape

(699, 11)

This dataframe contains 699 rows and 11 columns.

7. Display the first 5 rows of the dataframe using the method head():

In [7]:
df.head()

Sample code number  Clump Thickness  ...  Mitoses  Class
0             1000025                5  ...        1      2
1             1002945                5  ...        1      2
2             1015425                3  ...        1      2
3             1016277                6  ...        1      2
4             1017023                4  ...        1      2

[5 rows x 11 columns]

All the variables are numerical. The Sample code number column is an identifier for the measurement samples. 

8. Find the number of duplicated rows using the methods duplicated() and sum():

In [8]:
df.duplicated().sum()

8

Looking at the 11 columns in this dataset, there are 8 duplicated rows.

9. Display the duplicate rows using the methods loc() and duplicated():

In [9]:
df.loc[df.duplicated()]

Sample code number  Clump Thickness  ...  Mitoses  Class
208             1218860                1  ...        1      2
253             1100524                6  ...        3      4
254             1116116                9  ...        1      4
258             1198641                3  ...        1      2
272              320675                3  ...        1      4
338              704097                1  ...        1      2
561             1321942                5  ...        1      2
684              466906                1  ...        1      2

[8 rows x 11 columns]

The following rows are duplicates: 208, 253, 254, 258, 272, 338, 561 and 684.

10. Display the duplicate rows as for step 9 but with the parameter 'keep='last':

In [10]:
df.loc[df.duplicated(keep='last')]

Sample code number  Clump Thickness  ...  Mitoses  Class
42              1100524                6  ...        3      4
62              1116116                9  ...        1      4
168             1198641                3  ...        1      2
207             1218860                1  ...        1      2
267              320675                3  ...        1      4
314              704097                1  ...        1      2
560             1321942                5  ...        1      2
683              466906                1  ...        1      2

[8 rows x 11 columns]

With the parameter 'keep='last', the following rows are considered as duplicates: 42, 62, 168, 207, 267, 314, 560 and 683. Comparing this output to the one from the previous step, we can see rows 253 and 42 are identical.

11. Remove the duplicate rows using the method drop_duplicates() with the parameter keep='first' and saved it into a new dataframe called df_unique:

In [0]:
df_unique = df.drop_duplicates(keep='first')

12. Display the shape of df_unique with the attribute .shape:

In [12]:
df_unique.shape

(691, 11)

There are only 691 rows after removing the 8 duplicate records. Now this cleaned dataset contains only unique observations.

Excellent! In this exercise, we put in practise how to identify and remove duplicate records from a real world dataset.